File: src/main/java/com/redbox/domain/attach/strategy/FileAttachStrategyFactory.java
package com.redbox.domain.attach.strategy;

import com.redbox.domain.attach.entity.Category;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;

@Component
public class FileAttachStrategyFactory {

    private final Map<Category, FileAttachStrategy> strategies;

    public FileAttachStrategyFactory(List<FileAttachStrategy> strategyList) {
        strategies = Map.of(
                Category.NOTICE, findStrategy(strategyList, NoticeFileStrategy.class),
                Category.FUNDING, findStrategy(strategyList, RequestFileStrategy.class)
        );
    }

    // 전략 리스트에서 특정 타입의 전략을 찾는 private 메서드
    private FileAttachStrategy findStrategy(List<FileAttachStrategy> strategies,
                                            Class<? extends FileAttachStrategy> strategyClass) {
        return strategies.stream()
                .filter(strategyClass::isInstance)
                .findFirst()
                .orElseThrow(() -> new IllegalStateException("해당 전략이 설정되지 않았습니다: " + strategyClass.getSimpleName()));
    }

    public FileAttachStrategy getStrategy(Category category) {
        return strategies.get(category);
    }
}

File: migration/src/main/kotlin/com/redbox/domain/community/attach/strategy/FileAttachStrategyFactory.kt
package com.redbox.domain.community.attach.strategy

import com.redbox.domain.community.attach.entity.Category
import org.springframework.stereotype.Component

@Component
class FileAttachStrategyFactory(strategyList: List<FileAttachStrategy>) {

    // 전략을 저장하는 맵, Category를 키로 사용
    private val strategies: Map<Category, FileAttachStrategy> = mapOf(
        Category.FUNDING to findStrategy(strategyList, FundingFileStrategy::class.java),
        Category.NOTICE to findStrategy(strategyList, NoticeFileStrategy::class.java)
    )

    // 전략 리스트에서 특정 타입의 전략을 찾는 private 메서드
    private fun findStrategy(
        strategies: List<FileAttachStrategy>,
        strategyClass: Class<out FileAttachStrategy>
    ): FileAttachStrategy {
        return strategies.find { strategyClass.isInstance(it) }
            ?: throw IllegalStateException("해당 전략이 설정되지 않았습니다: ${strategyClass.simpleName}")
    }

    // 카테고리에 맞는 전략을 반환하는 메서드
    fun getStrategy(category: Category): FileAttachStrategy? {
        return strategies[category]
    }
}
